virtcontainers/persist/fs/fs.go

```
@@ -0,0 +1,262 @@
+func Name() string {
+	return "fs"
+}
+
+// Init FS persist driver and return abstract PersistDriver
+func Init() (persistapi.PersistDriver, error) {
+	return &FS{
+		sandboxState:   &persistapi.SandboxState{},
+		containerState: make(map[string]persistapi.ContainerState),
+		setFuncs:       make(map[string]persistapi.SetFunc),
+	}, nil
+}
+
+func (fs *FS) sandboxDir() (string, error) {
+	if fs.sandboxState.SandboxContainer == "" {
+		return "", fmt.Errorf("sandbox container id required")
+	}
+
+	return filepath.Join(runStoragePath, fs.sandboxState.SandboxContainer), nil
+}
+
+// ToDisk sandboxState and containerState to disk
+func (fs *FS) ToDisk() (retErr error) {
+	// call registered hooks to set sandboxState and containerState
+	for _, fun := range fs.setFuncs {
+		fun(fs.sandboxState, fs.containerState)
+	}
+
+	// if error happened, destroy all dirs
+	defer func() {
+		if retErr != nil {
+			// TODO: log error
+			fs.Destroy()
+		}
+	}()
+
+	sandboxDir, err := fs.sandboxDir()
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(sandboxDir, dirMode); err != nil {
+		return err
+	}
+
+	if err := fs.lock(); err != nil {
+		return err
+	}
+	defer fs.unlock()
+
+	// persist sandbox configuration data
+	sandboxFile := filepath.Join(sandboxDir, persistFile)
+	f, err := os.OpenFile(sandboxFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileMode)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(fs.sandboxState); err != nil {
+		return err
+	}
+
+	// persist container configuration data
+	for cid, cstate := range fs.containerState {
+		cdir := filepath.Join(sandboxDir, cid)
+		if err := os.MkdirAll(cdir, dirMode); err != nil {
+			return err
+		}
+
+		cfile := filepath.Join(cdir, persistFile)
+		cf, err := os.OpenFile(cfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileMode)
+		if err != nil {
+			return err
+		}
+
+		if err := json.NewEncoder(cf).Encode(cstate); err != nil {
+			return err
+		}
+		cf.Close()
```

handle/log error if cf.Close() returns one?

---

I think this is a normal handling flow 'defer f.Close()', ignoring the error is safe.